frame/staking/src/ledger.rs

```diff
@@ -0,0 +1,419 @@
+		}
+	}
+
+	/// Returns a staking ledger, if it is bonded and it exists in storage.
+	///
+	/// This getter can be called with either a controller or stash account, provided that the
+	/// account is properly wrapped in the respective [`StakingAccount`] variant. This is meant to
+	/// abstract the concept of controller/stash accounts to the caller.
+	pub(crate) fn get(account: StakingAccount<T::AccountId>) -> Result<StakingLedger<T>, Error<T>> {
+		let controller = match account {
+			StakingAccount::Stash(stash) => <Bonded<T>>::get(stash).ok_or(Error::<T>::NotStash),
+			StakingAccount::Controller(controller) => Ok(controller),
+		}?;
+
+		<Ledger<T>>::get(&controller)
+			.map(|mut ledger| {
+				ledger.controller = Some(controller.clone());
+				ledger
+			})
+			.ok_or_else(|| Error::<T>::NotController)
+	}
+
+	/// Returns the controller account of a staking ledger.
+	///
+	/// Note: it will fallback into querying the `Bonded` storage with the ledger stash if the
+	/// controller is not set in `self`, which most likely means that self was fetched directly from
+	/// `Ledger` instead of through the methods exposed in `StakingLedger`. If the ledger does not
+	/// exist in storage, it returns `None`.
+	pub(crate) fn controller(&self) -> Option<T::AccountId> {
+		self.controller
+			.clone()
+			.or_else(|| Self::paired_account(StakingAccount::Stash(self.stash.clone())))
+	}
+
+	/// Inserts/updates a staking ledger account.
+	///
+	/// Bonds the ledger if it is not bonded yet, signalling that this is a new ledger. The staking
+	/// locks of the stash account are updated accordingly.
+	///
+	/// Note: To ensure lock consistency, all the [`Ledger`] storage updates should be made through
+	/// this helper function.
+	pub(crate) fn update(&self) -> Result<(), Error<T>> {
+		if !<Bonded<T>>::contains_key(&self.stash) {
+			// not bonded yet, new ledger. Note: controllers are deprecated, stash is the
+			// controller.
+			<Bonded<T>>::insert(&self.stash, &self.stash);
+		}
+
+		T::Currency::set_lock(STAKING_ID, &self.stash, self.total, WithdrawReasons::all());
+		Ledger::<T>::insert(&self.controller().ok_or(Error::<T>::NotController)?, &self);
+
+		Ok(())
+	}
+
+	// Bonds a ledger.
+	//
+	// This method is just syntactic sugar of [`Self::update`] with a check that returns an error if
+	// the ledger has is already bonded to ensure that the method behaves as expected.
+	pub(crate) fn bond(&self) -> Result<(), Error<T>> {
+		if <Bonded<T>>::contains_key(&self.stash) {
+			Err(Error::<T>::AlreadyBonded)
+		} else {
+			self.update()
+		}
+	}
+
+	/// Clears all data related to a staking ledger and its bond in both [`Ledger`] and [`Bonded`]
+	/// storage items and updates the stash staking lock.
+	pub(crate) fn kill(stash: &T::AccountId) -> Result<(), Error<T>> {
+		let controller = <Bonded<T>>::get(stash).ok_or(Error::<T>::NotStash)?;
+
+		<Ledger<T>>::get(&controller).ok_or(Error::<T>::NotController).map(|ledger| {
```

Comment:
Should we consolidate the type syntax to turbofish syntax? This would make pallet-wide searches of storage quicker.

```suggestion
		Ledger::<T>::get(&controller).ok_or(Error::<T>::NotController).map(|ledger| {
```